File: examples/language-impl-comparison/ocaml/pl0.ml
type op = Add | Sub | Mul | Div

type expr =
  | Number of int
  | Var of string
  | Binary of op * expr * expr

type stmt =
  | Assign of string * expr
  | While of expr * stmt list
  | Write of expr

exception Exec_error of string

type runtime = { vars : (string, int) Hashtbl.t; output : int Queue.t }

let create_runtime () = { vars = Hashtbl.create 16; output = Queue.create () }

let rec eval_expr rt = function
  | Number n -> n
  | Var name -> (try Hashtbl.find rt.vars name with Not_found -> raise (Exec_error ("未定義変数: " ^ name)))
  | Binary (op, lhs, rhs) ->
      let l = eval_expr rt lhs in
      let r = eval_expr rt rhs in
      match op with
      | Add -> l + r
      | Sub -> l - r
      | Mul -> l * r
      | Div -> if r = 0 then raise (Exec_error "0 で割れません") else l / r

let rec exec_stmt rt = function
  | Assign (name, expr) ->
      let value = eval_expr rt expr in
      Hashtbl.replace rt.vars name value
  | Write expr ->
      let value = eval_expr rt expr in
      Queue.add value rt.output
  | While (cond, body) ->
      while eval_expr rt cond <> 0 do
        List.iter (exec_stmt rt) body
      done

let exec program =
  let rt = create_runtime () in
  List.iter (exec_stmt rt) program;
  rt

File: examples/language-impl-comparison/rust/pl0.rs
use std::collections::HashMap;

#[derive(Debug, Clone)]
pub enum Stmt {
    Assign { name: String, expr: Expr },
    While { cond: Expr, body: Vec<Stmt> },
    Write { expr: Expr },
}

#[derive(Debug, Clone)]
pub enum Expr {
    Number(i64),
    Var(String),
    Binary { op: Op, lhs: Box<Expr>, rhs: Box<Expr> },
}

#[derive(Debug, Clone, Copy)]
pub enum Op {
    Add,
    Sub,
    Mul,
    Div,
}

#[derive(Default)]
pub struct Runtime {
    vars: HashMap<String, i64>,
    pub output: Vec<i64>,
}

#[derive(Debug)]
pub enum ExecError {
    UndefinedVar(String),
    DivisionByZero,
}

pub fn exec(program: &[Stmt]) -> Result<Runtime, ExecError> {
    let mut runtime = Runtime::default();
    for stmt in program {
        exec_stmt(stmt, &mut runtime)?;
    }
    Ok(runtime)
}

fn exec_stmt(stmt: &Stmt, runtime: &mut Runtime) -> Result<(), ExecError> {
    match stmt {
        Stmt::Assign { name, expr } => {
            let value = eval_expr(expr, &runtime.vars)?;
            runtime.vars.insert(name.clone(), value);
        }
        Stmt::While { cond, body } => {
            while eval_expr(cond, &runtime.vars)? != 0 {
                for inner in body {
                    exec_stmt(inner, runtime)?;
                }
            }
        }
        Stmt::Write { expr } => {
            let value = eval_expr(expr, &runtime.vars)?;
            runtime.output.push(value);
        }
    }
    Ok(())
}

fn eval_expr(expr: &Expr, vars: &HashMap<String, i64>) -> Result<i64, ExecError> {
    match expr {
        Expr::Number(n) => Ok(*n),
        Expr::Var(name) => vars
            .get(name)
            .copied()
            .ok_or_else(|| ExecError::UndefinedVar(name.clone())),
        Expr::Binary { op, lhs, rhs } => {
            let l = eval_expr(lhs, vars)?;
            let r = eval_expr(rhs, vars)?;
            Ok(match op {
                Op::Add => l + r,
                Op::Sub => l - r,
                Op::Mul => l * r,
                Op::Div => {
                    if r == 0 {
                        return Err(ExecError::DivisionByZero);
                    }
                    l / r
                }
            })
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn run_simple_loop() {
        let program = vec![
            Stmt::Assign { name: "x".into(), expr: Expr::Number(3) },
            Stmt::While {
                cond: Expr::Var("x".into()),
                body: vec![
                    Stmt::Write { expr: Expr::Var("x".into()) },
                    Stmt::Assign {
                        name: "x".into(),
                        expr: Expr::Binary {
                            op: Op::Sub,
                            lhs: Box::new(Expr::Var("x".into())),
                            rhs: Box::new(Expr::Number(1)),
                        },
                    },
                ],
            },
        ];

        let runtime = exec(&program).unwrap();
        assert_eq!(runtime.output, vec![3, 2, 1]);
    }
}
